--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +_,14 @@
      */
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
+
+    // DivineMC start
+    /**
+     * The brand id for DivineMC.
+     */
+    Key BRAND_DIVINEMC_ID = Key.key("divinemc", "divinemc");
+    // DivineMC end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
