--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -320,6 +_,12 @@
                 if (!this.isSilent()) {
                     serverLevel.levelEvent(null, 1027, this.blockPosition(), 0);
                 }
+
+            // DivineMC start - Fix MC-200418
+            if (villager.isPassenger() && villager.getVehicle() instanceof net.minecraft.world.entity.animal.Chicken && villager.isBaby()) {
+                villager.removeVehicle();
+            }
+            // DivineMC end - Fix MC-200418
                 // CraftBukkit start
             }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.CURED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CURED // CraftBukkit
         );
