--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -7,7 +_,8 @@
 import net.minecraft.util.LinearCongruentialGenerator;
 import net.minecraft.util.Mth;
 
-public class BiomeManager {
+public class
+BiomeManager {
     public static final int CHUNK_CENTER_QUART = QuartPos.fromBlock(8);
     private static final int ZOOM_BITS = 2;
     private static final int ZOOM = 4;
