--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -6632,6 +_,7 @@
             .mapColor(MapColor.COLOR_ORANGE)
             .instrument(NoteBlockInstrument.BASEDRUM)
             .requiresCorrectToolForDrops()
+            .sound(SoundType.COPPER) // DivineMC - Fix MC-223153
             .strength(5.0F, 6.0F)
     );
     public static final Block RAW_GOLD_BLOCK = register(
