--- a/net/minecraft/server/commands/LocateCommand.java
+++ b/net/minecraft/server/commands/LocateCommand.java
@@ -200,8 +_,10 @@
     }
 
     private static float dist(int x1, int z1, int x2, int z2) {
-        int i = x2 - x1;
-        int i1 = z2 - z1;
-        return Mth.sqrt(i * i + i1 * i1);
+        // DivineMC start - Fix MC-177381
+        double i = x2 - x1;
+        double j = z2 - z1;
+        return (float) Math.hypot(i, j);
+        // DivineMC end
     }
 }
