--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -344,7 +_,7 @@
         return this.isInGround() && this.level().noCollision(new AABB(this.position(), this.position()).inflate(0.06));
     }
 
-    private void startFalling() {
+    public void startFalling() { // DivineMC - private -> public - AbstractArrow#startFalling
         this.setInGround(false);
         Vec3 deltaMovement = this.getDeltaMovement();
         this.setDeltaMovement(deltaMovement.multiply(this.random.nextFloat() * 0.2F, this.random.nextFloat() * 0.2F, this.random.nextFloat() * 0.2F));
