--- a/purpur-api/build.gradle.kts
+++ b/purpur-api/build.gradle.kts
@@ -104,17 +_,21 @@
         java {
             srcDir(generatedApiPath)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../purpur-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../purpur-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../purpur-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../purpur-api/src/test/resources"))
         }
     }
 }
@@ -161,6 +_,15 @@
         pomProps.get().asFile.writeText("version=$projectVersion")
     }
 }
+
+// DivineMC start - Hide unnecessary compilation warnings
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("-Xlint:-module")
+    compilerArgs.add("-Xlint:-removal")
+    compilerArgs.add("-Xlint:-dep-ann")
+}
+// DivineMC end - Hide unnecessary compilation warnings
 
 tasks.jar {
     from(generateApiVersioningFile.map { it.outputs.files.singleFile }) {
