From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Mon, 27 Mar 2023 17:50:05 +0300
Subject: [PATCH] Divine Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index c293dbc96051a5a5d11dc445b8c389e2ce33ec84..19df56efa6a6911cd5eda620dc20082885317954 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1983,6 +1983,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // DivineMC start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDivineConfig() {
+            throw new UnsupportedOperationException("Not supported yet");
+        }
+        // DivineMC end
+
         /**
          * Sends the component to the player
          *
