From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Mon, 27 Mar 2023 17:50:05 +0300
Subject: [PATCH] Divine Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 15a0ae9744fa841e3e51e9e9a957cea8c71e68ad..aae4102b23e455516ddd80ed8f5ebd62b3e5d519 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2019,6 +2019,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // DivineMC start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDivineConfig() {
+            throw new UnsupportedOperationException("Not supported yet");
+        }
+        // DivineMC end
+
         /**
          * Sends the component to the player
          *
