From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Mon, 27 Mar 2023 17:50:05 +0300
Subject: [PATCH] Divine Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ec8b9b2cbe65838a194281f7d76d0e17defc5211..adeb76886e99bd8f8374102471c71694ab5a30e4 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2269,6 +2269,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // DivineMC start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDivineConfig() {
+            throw new UnsupportedOperationException("Not supported yet");
+        }
+        // DivineMC end
+
         /**
          * Sends the component to the player
          *
