From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Mon, 27 Mar 2023 17:50:05 +0300
Subject: [PATCH] Divine Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ceb81148f5fd17f00d1a0fb2df1f31c45aa6807d..a27599cb3f3274fe0a4bf09b6477d93440c28bce 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2263,6 +2263,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // DivineMC start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getDivineConfig() {
+            throw new UnsupportedOperationException("Not supported yet");
+        }
+        // DivineMC end
+
         /**
          * Sends the component to the player
          *
