From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Tue, 14 Mar 2023 21:03:19 +0300
Subject: [PATCH] Add packages


diff --git a/build.gradle.kts b/build.gradle.kts
index 181e9cd8623995f40e696ccfe49754dc340405d8..919aaa4821e65cd2d6ee39b4740d9bfadfb7c058 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -152,3 +152,23 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+
+// DivineMC start
+publishing {
+    repositories {
+        maven {
+            name = "githubPackage"
+            url = uri("https://maven.pkg.github.com/DivineMC/DivineMC")
+
+            credentials.username = System.getenv("GITHUB_USERNAME")
+            credentials.password = System.getenv("GITHUB_TOKEN")
+        }
+
+        publications {
+            register<MavenPublication>("gpr") {
+                from(components["java"])
+            }
+        }
+    }
+}
+// DivineMC end
\ No newline at end of file
