From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sat, 10 Jun 2023 13:17:01 +0300
Subject: [PATCH] Fix MC-31819


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index dfc73cd1b4fd9a2572a71b653e6e3d9133ec634c..2726f34d88aecf92f90396f760f6a333fbe5bebf 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1905,6 +1905,11 @@ public abstract class Player extends LivingEntity {
     }
 
     public void causeFoodExhaustion(float f, EntityExhaustionEvent.ExhaustionReason reason) {
+        // DivineMC start - Fix MC-31819
+        if (this.level().getDifficulty() == Difficulty.PEACEFUL) {
+            return;
+        }
+        // DivineMC end
         // CraftBukkit end
         if (!this.abilities.invulnerable) {
             if (!this.level().isClientSide) {
