From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 16 Jul 2023 11:37:32 +0300
Subject: [PATCH] Make entity goals public


diff --git a/src/main/java/net/minecraft/world/entity/animal/Bee.java b/src/main/java/net/minecraft/world/entity/animal/Bee.java
index bdb78cc701543cfe91a6bafd1786fe2ea0bf1ddc..40c31cc3a83a8776d396d5d14fce8c795ab08bd3 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Bee.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Bee.java
@@ -774,7 +774,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         return pos.closerThan(this.blockPosition(), (double) distance);
     }
 
-    private class BeePollinateGoal extends Bee.BaseBeeGoal {
+    public class BeePollinateGoal extends Bee.BaseBeeGoal { // DivineMC - private -> public
 
         private static final int MIN_POLLINATION_TICKS = 400;
         private static final int MIN_FIND_FLOWER_RETRY_COOLDOWN = 20;
@@ -963,7 +963,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private class BeeLookControl extends org.purpurmc.purpur.controller.LookControllerWASD { // Purpur
+    public class BeeLookControl extends org.purpurmc.purpur.controller.LookControllerWASD { // Purpur // DivineMC - private -> public
 
         BeeLookControl(Mob entity) {
             super(entity);
@@ -999,7 +999,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private class BeeEnterHiveGoal extends Bee.BaseBeeGoal {
+    public class BeeEnterHiveGoal extends Bee.BaseBeeGoal { // DivineMC - private -> public
 
         BeeEnterHiveGoal() {
             super();
@@ -1044,7 +1044,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private class BeeLocateHiveGoal extends Bee.BaseBeeGoal {
+    public class BeeLocateHiveGoal extends Bee.BaseBeeGoal { // DivineMC - private -> public
 
         BeeLocateHiveGoal() {
             super();
@@ -1276,7 +1276,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private class BeeGrowCropGoal extends Bee.BaseBeeGoal {
+    public class BeeGrowCropGoal extends Bee.BaseBeeGoal { // DivineMC - private -> public
 
         static final int GROW_CHANCE = 30;
 
@@ -1340,7 +1340,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private class BeeWanderGoal extends Goal {
+    public class BeeWanderGoal extends Goal { // DivineMC - private -> public
 
         private static final int WANDER_THRESHOLD = 22;
 
@@ -1387,7 +1387,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private class BeeHurtByOtherGoal extends HurtByTargetGoal {
+    public class BeeHurtByOtherGoal extends HurtByTargetGoal { // DivineMC - private -> public
 
         BeeHurtByOtherGoal(Bee entitybee) {
             super(entitybee);
@@ -1407,7 +1407,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private static class BeeBecomeAngryTargetGoal extends NearestAttackableTargetGoal<Player> {
+    public static class BeeBecomeAngryTargetGoal extends NearestAttackableTargetGoal<Player> { // DivineMC - private -> public
 
         BeeBecomeAngryTargetGoal(Bee bee) {
             // Objects.requireNonNull(entitybee); // CraftBukkit - decompile error
@@ -1438,7 +1438,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
         }
     }
 
-    private abstract class BaseBeeGoal extends Goal {
+    public abstract class BaseBeeGoal extends Goal { // DivineMC - private -> public
 
         BaseBeeGoal() {}
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cat.java b/src/main/java/net/minecraft/world/entity/animal/Cat.java
index 6e5b4a50b8d2dddbd7f099426b4866b35bcb798a..2bcd475cca90d356de056f9b8f023d0d6392fe44 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cat.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cat.java
@@ -568,7 +568,7 @@ public class Cat extends TamableAnimal implements VariantHolder<CatVariant> {
         }
     }
 
-    private static class CatRelaxOnOwnerGoal extends Goal {
+    public static class CatRelaxOnOwnerGoal extends Goal { // DivineMC - private -> public
 
         private final Cat cat;
         @Nullable
@@ -714,7 +714,7 @@ public class Cat extends TamableAnimal implements VariantHolder<CatVariant> {
         }
     }
 
-    private static class CatAvoidEntityGoal<T extends LivingEntity> extends AvoidEntityGoal<T> {
+    public static class CatAvoidEntityGoal<T extends LivingEntity> extends AvoidEntityGoal<T> { // DivineMC - private -> public
 
         private final Cat cat;
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/Pillager.java b/src/main/java/net/minecraft/world/entity/monster/Pillager.java
index 06d52d8b61abc4dbbdc953bfed2e688be377b3cc..a5490068d200ed78ca720ae804a1b26faa4cba7f 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Pillager.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Pillager.java
@@ -99,7 +99,7 @@ public class Pillager extends AbstractIllager implements CrossbowAttackMob, Inve
         super.registerGoals();
         this.goalSelector.addGoal(0, new FloatGoal(this));
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur
-        this.goalSelector.addGoal(2, new Raider.HoldGroundAttackGoal(this, 10.0F));
+        this.goalSelector.addGoal(2, new HoldGroundAttackGoal(this, 10.0F));
         this.goalSelector.addGoal(3, new RangedCrossbowAttackGoal<>(this, 1.0D, 8.0F));
         this.goalSelector.addGoal(8, new RandomStrollGoal(this, 0.6D));
         this.goalSelector.addGoal(9, new LookAtPlayerGoal(this, Player.class, 15.0F, 1.0F));
diff --git a/src/main/java/net/minecraft/world/entity/monster/Vindicator.java b/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
index ffe956f200dd0b4148043d0d7f2eac007a9f22b1..a5608ddd34afd96ac166758f5d8ec88ed92d34cd 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Vindicator.java
@@ -97,7 +97,7 @@ public class Vindicator extends AbstractIllager {
         this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur
         this.goalSelector.addGoal(1, new Vindicator.VindicatorBreakDoorGoal(this));
         this.goalSelector.addGoal(2, new AbstractIllager.RaiderOpenDoorGoal(this));
-        this.goalSelector.addGoal(3, new Raider.HoldGroundAttackGoal(this, 10.0F));
+        this.goalSelector.addGoal(3, new HoldGroundAttackGoal(this, 10.0F));
         this.goalSelector.addGoal(4, new MeleeAttackGoal(this, 1.0D, false));
         this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur
         this.targetSelector.addGoal(1, (new HurtByTargetGoal(this, Raider.class)).setAlertOthers());
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raider.java b/src/main/java/net/minecraft/world/entity/raid/Raider.java
index 300411e88cdaef7e0ce280edbd5f18fa8d004a33..80e0a66abd9cd6ad9135658adb3044f97e0b7ba1 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raider.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raider.java
@@ -352,7 +352,7 @@ public abstract class Raider extends PatrollingMonster {
         }
     }
 
-    private static class RaiderMoveThroughVillageGoal extends Goal {
+     public class RaiderMoveThroughVillageGoal extends Goal { // DivineMC - private -> public
 
         private final Raider raider;
         private final double speedModifier;
@@ -500,7 +500,7 @@ public abstract class Raider extends PatrollingMonster {
         }
     }
 
-    protected class HoldGroundAttackGoal extends Goal {
+    public static class HoldGroundAttackGoal extends Goal { // DivineMC - protected -> public static
 
         private final Raider mob;
         private final float hostileRadiusSqr;
