From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Wed, 22 May 2024 23:32:09 +0300
Subject: [PATCH] Fix possible NPE during creating GUI graph


diff --git a/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java b/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java
index 308f67d0616e2d6bb135258f1fda53ccdee01430..553528e612b72199ac33404c086db0e3b6887e7e 100644
--- a/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java
+++ b/src/main/java/com/destroystokyo/paper/gui/RAMDetails.java
@@ -59,7 +59,13 @@ public class RAMDetails extends JList<String> {
         GraphData data = RAMGraph.DATA.peekLast();
         Vector<String> vector = new Vector<>();
 
-        double[] tps = org.bukkit.Bukkit.getTPS();
+        // DivineMC start - Fix possible NPE during creating GUI graph
+        double[] tps = new double[]{
+                server.tps1.getAverage(),
+                server.tps5.getAverage(),
+                server.tps15.getAverage()
+        };
+        // DivineMC end
         String[] tpsAvg = new String[tps.length];
 
         for ( int g = 0; g < tps.length; g++) {
